objects/isn_unifiedgrowingtray.lua

```diff
@@ -0,0 +1,511 @@
+			storage.stages = index
+			local stage = {}
+			stage.min = curStage.duration[1] or 10
+			stage.max = curStage.duration[2] or stage.min
+			storage.stage[index] = stage
+		else
+			--We expect this branch to only happen once but just in case we use 'or'.
+			storage.harvestPool = curStage.harvestPool or storage.harvestPool
+			storage.resetStage = (curStage.resetToStage and curStage.resetToStage + 1) or storage.resetStage
+			if curStage.tree then
+				if not isn_readSaplingData() then return false end
+				storage.harvestPool = ""
+			end
+		end
+		index = index + 1
+	end
+	if index == 1 then return false end
+	return true
+end
+
+--Reads sapling data and does minimal verification that we indeed have a valid
+--sapling.
+function isn_readSaplingData()
+	if type(storage.currentseed.parameters) ~= "table" then return false end
+	local stemType = storage.currentseed.parameters.stemName
+	local foliageType = storage.currentseed.parameters.foliageName
+	if not (stemType and foliageType) then return false end
+	local stemPath = root.treeStemDirectory(stemType)
+	local stemJson = root.assetJson(stemPath .. stemType .. ".modularstem")
+	storage.stemDrops = stemJson.dropConfig.drops
+	local foliagePath = root.treeFoliageDirectory(foliageType)
+	local foliageJson = root.assetJson(foliagePath .. foliageType .. ".modularfoliage")
+	storage.foliageDrops = foliageJson.dropConfig.drops
+	storage.hasTree = true
+	return true
+end
+
+--Generates drop data for a sapling
+--[[
+If I follow correctly, the stem and the foliage each can have a single item type drop.
+That drop is randomly picked from "dropConfig/drops" table.
+Equal weight, each drop line is a single item type and coresponding count.
+Also foliage is the only source for saplings.
+]]--
+function isn_genSaplingDrops()
+	local drops = {}
+	local dropCount = 1
+	if type(storage.stemDrops) == "table" then
+		drops[dropCount] = root.createItem(storage.stemDrops[math.random(1, #storage.stemDrops)][1], world.threatLevel())
+		dropCount = dropCount + 1
+	end
+	if type(storage.foliageDrops) == "table" then
+		drops[dropCount] = root.createItem(storage.foliageDrops[math.random(1, #storage.foliageDrops)][1], world.threatLevel())
+		if drops[dropCount].name == "sapling" then drops[dropCount].parameters = storage.currentseed.parameters end
+		dropCount = dropCount + 1
+	end
+	if dropCount == 1 then return end
+	return drops
+end
+
+--Generates growth data to tell when a plant is ready for harvest and when it
+--needs to be watered.
+function isn_genGrowthData(initStage)
+	initStage = initStage or 1
+	--Make sure some things make sense for a perennial, if not act like it's annual
+	if initStage > storage.stages then
+		initStage = 1
+		storage.resetStage = 1
+	end
+	storage.currentStage = initStage
+	storage.growthCap = 0
+	local index = initStage
+	while index <= storage.stages do
+		local stage = storage.stage[index]
+		if type(stage) == nil then return false end
```

Comment:
`if not type(stage) then return end`

Reply:
Similar thing here... we are intending to return a boolean.
With what you are suggesting (for the return type) means that the following lines of code are no longer equivalent when they should be.
`if isn_genGrowthData() == false` then ... end`
`if not isn_genGrowthData() then ... end`
Proper behaviour in this case is since we are intending to return a boolean, explicitly return the false and not let it go nil.  This is because nil ~= false.